### ISSUE TITLE
Chunked inference

### ISSUE THREAD

**Issue by yhv-wt** [state-spaces/mamba#536]
Hi,

First thank you for your incredible work!

From what I can tell from the code there are two paths for inference - either by sending the complete input sequence using the fast path when `inference_params is None` or much slower sending one input (token or whatever) at a time, when `inference_params.seqlen_offset > 0`. This design mostly caters for chats, i.e. process prompt fast using triton and then generate response slowly token-by-token. But not every AI problem is a chat. It's still unclear to me what the third path is for, i.e. when `inference_params is not None` but `inference_params.seqlen_offset == 0` but doesn't seem it's for what I'm asking.
So my question is there any way to implement fast chunked inference of potentially very long sequences (that otherwise won't fit into memory) by passing final SSM and conv states between chunks and still using fast triton kernels path or is there some fundamental restriction of this architecture that only allows continued inference one token at a time?

Thanks!

**Comment by OleguerCanal** [comment]
I also have the same doubt. Just to clarify:

We'd like `step` function in the Mamba modules to be able to take `t > 1` tokens. As of now there is this assert:

```python
assert hidden_states.shape[1] == 1, "Only support decoding with 1 token at a time for now"
```

Are there plans to address this case? Is it very hard to do @tridao? Thanks!

**Comment by assafbk** [comment]
+1, it will be great if chunked parallel mode will be supported (even if just for inference).

Just to clarify a few things:
@yhv-wt the first path (inference_params is None) is for training in parallel mode.
The second path is for inference in sequential mode.
The third path is for inference in parallel mode.
The two latter paths are an efficient way to perform inference - process the context in parallel mode and then predict new tokens in recurrent mode.

@OleguerCanal Not sure if I fully understand your request - but sequential inference mode ('step') predicts one token at a time. Since each token depends on the previous tokens, I dont see how this can be sped up.

**Comment by OleguerCanal** [comment]
@assafbk only if you are running a model in an auto-regressive manner (like in next-token prediction applications such as language models). Some applications that do sequence-to-sequence have available all input tokens from the beginning even at inference time. It is just too much memory to run it all at once, that's why I want to chunk it.

Not sure I am expressing myself correctly.

**Comment by assafbk** [comment]
Right. I think that we are talking about the same thing.
I just highlighted that what you are asking for is not possible via 'step' (the third path). It is possible via the second path (inference in parallel mode).

**Comment by 843773493** [comment]
May be like this?
`        import easydict
        batch, length, dim, chunk_size = 2, 64, 64, 4
        x = torch.randn(batch, length, dim).to("cuda")
        model = Mamba2(
            # This module uses roughly 3 * expand * d_model^2 parameters
            d_model=dim, # Model dimension d_model
            d_state=64,  # SSM state expansion factor, typically 64 or 128
            d_conv=4,    # Local convolution width
            expand=2,    # Block expansion factor
            headdim=dim//8,
            layer_idx='0', # Layer index for inference cache
        ).to("cuda")
        inference_params = easydict.EasyDict()
        inference_params.key_value_memory_dict = {'0':model.allocate_inference_cache(batch, max_seqlen=None)}
        inference_params.seqlen_offset = 0
        for i in range(chunk_size):
            y_ = model(x[:, i*(length//chunk_size):(i+1)*(length//chunk_size)], inference_params=inference_params)
            print(
                y_.mean().item(), 
                inference_params.key_value_memory_dict['0'][0].shape,
                inference_params.key_value_memory_dict['0'][0].mean().item(), 
                inference_params.key_value_memory_dict['0'][1].shape,
                inference_params.key_value_memory_dict['0'][1].mean().item(),
            )`

**Comment by 843773493** [comment]
May be like this?
```
import easydict
batch, length, dim, chunk_size = 2, 64, 64, 4
x = torch.randn(batch, length, dim).to("cuda")
model = Mamba2(
    # This module uses roughly 3 * expand * d_model^2 parameters
    d_model=dim, # Model dimension d_model
    d_state=64,  # SSM state expansion factor, typically 64 or 128
    d_conv=4,    # Local convolution width
    expand=2,    # Block expansion factor
    headdim=dim//8,
    layer_idx='0', # Layer index for inference cache
).to("cuda")
inference_params = easydict.EasyDict()
inference_params.key_value_memory_dict = {'0':model.allocate_inference_cache(batch, max_seqlen=None)}
inference_params.seqlen_offset = 0
for i in range(chunk_size):
    y_ = model(x[:, i*(length//chunk_size):(i+1)*(length//chunk_size)], inference_params=inference_params)
    print(
        y_.mean().item(), 
        inference_params.key_value_memory_dict['0'][0].shape,
        inference_params.key_value_memory_dict['0'][0].mean().item(), 
        inference_params.key_value_memory_dict['0'][1].shape,
        inference_params.key_value_memory_dict['0'][1].mean().item(),
    )
```

**Comment by wren93** [comment]
> Right. I think that we are talking about the same thing. I just highlighted that what you are asking for is not possible via 'step' (the third path). It is possible via the second path (inference in parallel mode).

@assafbk @OleguerCanal  Hello, may I ask if there is an off-the-shelf inference setting that supports the second path you mentioned? I have a similar problem where I have a long sequence of input, I just want to collect the output for every token but I don't need to autoregressively predict new tokens. I don't really care about runtime so I want to process these tokens in sequential mode during inference to save memory. Is there a way to do this? Thanks

**Comment by assafbk** [comment]
I would try something like this:

```
with torch.inference_mode():
    output = model(input_ids, num_last_tokens=input_ids.shape[1])
    logits = output.logits
    predictions = torch.max(logits, dim=2)[0]
```

**Comment by rschiewer** [comment]
Sorry for reactivating this discussion, but just to clarify since the issue is still open: Chunked inference seems to be supported with this Mamba implementation, according to what I understood. Is that correct? If so, does it also work during training?

**I'm basically looking for a mamba implementation that provides the same interface as say a GRU layer from PyTorch.** Specifically, you when invoking the model you can provide a hidden state or cache from a previous invocation if you want. Furthermore, after model invocation you get the most recent hidden state or cache and the output from the model. At the same time, the implementation uses the efficient and fast implementation for all of this.

**A bit of context why I'm looking for this functionality:** I'm using Mamba in a setting where during training I need to process sequences in chunks. I apply additional computation to each chunk and based on the result choose the next one, which makes processing the whole sequence at once impossible. I could just use the step() method for this, but it's very slow. Thus, I'm looking for a way to process sequence parts in a fast an efficient way while still being able to maintain the cache throughout.

**Comment by mathieu-charbonnel** [comment]
[Solution for chunked Inference]

Hi,

At Datadog we have been using a Mamba 1 model to scan very long streams of (tokenized) logs chunk by chunk. We evaluate the log-likelihood of each log. Given the current code only the first chunk can be processed in parallel. Once we want to use the state and convolution inputs, we unfortunately have to switch to token by token prediction. 

As Yaroslav mentions, in a context like this that is not the classic autoregressive decoding, chunked processing would be a direct win for us as we do not need next-token feedback at each step to compute per-token likelihood. We want to benefit from the parallelism of pscan for inference too, while also feeding convolution inputs and the state.

Generally speaking I believe that enabling a chunk by chunk inference is reasonably easy and would make a nice new feature to the library.
Even for a more classic autoregressive decoding usecase, processing prompts that have state with parallel scan could make every user prompt after the first substantially faster to process.

At Datadog I worked on a fix internally and I would like to share it with the community, I will open a PR that adds this chunked_inference feature.

**Comment by oniwakaa** [comment]
placeholder to get comments
